optimizely/event/event_factory.py

```diff
@@ -0,0 +1,181 @@
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+
+from .user_event import ConversionEvent, ImpressionEvent
+from .payload import Decision, EventBatch, Snapshot, SnapshotEvent, Visitor, VisitorAttribute
+from .log_event import LogEvent
+from optimizely.helpers import enums, event_tag_utils, validator
+
+CUSTOM_ATTRIBUTE_FEATURE_TYPE = 'custom'
+
+
+class EventFactory(object):
+  """ EventFactory builds LogEvent object from a given UserEvent.
+  This class serves to separate concerns between events in the SDK and the API used
+  to record the events via the Optimizely Events API ("https://developers.optimizely.com/x/events/api/index.html")
+  """
+
+  EVENT_ENDPOINT = 'https://logx.optimizely.com/v1/events'
+  HTTP_VERB = 'POST'
+  HTTP_HEADERS = {'Content-Type': 'application/json'}
+  ACTIVATE_EVENT_KEY = 'campaign_activated'
+
+  @classmethod
+  def create_log_event(cls, user_events, logger):
+    """ Create LogEvent instance.
+
+    Args:
+      user_events: An array of UserEvent instances.
+      logger: Provides a logger instance.
+
+    Returns:
+      LogEvent instance.
+    """
+
+    def _dict_clean(obj):
+      """ Helper method to remove keys from dictionary with None values. """
+
+      result = {}
+      for k, v in obj:
+        if v is None and k in ['revenue', 'value', 'tags', 'decisions']:
+          continue
+        else:
+          result[k] = v
+      return result
+
+    if not isinstance(user_events, list):
+      user_events = [user_events]
+
+    visitors = []
+
+    for user_event in user_events:
+      visitors.append(cls._create_visitor(user_event, logger))
+      user_context = user_event.event_context
+
+      event_batch = EventBatch(
+        user_context.account_id,
+        user_context.project_id,
+        user_context.revision,
+        user_context.client_name,
+        user_context.client_version,
+        user_context.anonymize_ip,
+        True
+      )
+
+    if len([x for x in visitors if x is not None]) == 0:
+      return None
+
+    event_batch.visitors = visitors
+
+    event_params = json.loads(
+        json.dumps(event_batch.__dict__, default=lambda o: o.__dict__),
+        object_pairs_hook=_dict_clean
+      )
+
+    return LogEvent(cls.EVENT_ENDPOINT, event_params, cls.HTTP_VERB, cls.HTTP_HEADERS)
+
+  @classmethod
+  def _create_visitor(cls, user_event, logger):
+    if not user_event:
+      return None
+
+    if isinstance(user_event, ImpressionEvent):
+      decision = Decision(
+        user_event.experiment.layerId if user_event.experiment else None,
```

Comment:
If `user_event.experiment` is `None` or `user_event.variation` is `None`, that seems like an error or invalid data, right? Shouldn't all `ImpressionEvents` have `experiment` and `variation`?